File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        
        def sums(root, curr_val):
            nonlocal ans
            if(root):
                val = (curr_val << 1) | root.val
                if(not root.left and not root.right):
                    ans += val
                    return
                
                sums(root.left, val)
                sums(root.right, val)
        
        ans = 0
        sums(root, 0)
        return ans
        

File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumRootToLeaf(root *TreeNode) int {
    tot := 0
    pathSum(root, 0, &tot)
    return tot
}

func pathSum(node *TreeNode, val int, tot *int) {
    if node == nil {
        return
    }

    val = (val << 1) + node.Val
    if node.Left == nil && node.Right == nil {
        *tot += val
        return
    }
    pathSum(node.Left, val, tot)
    pathSum(node.Right, val, tot)
}